ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureBlobDataAccessor.java

```diff
@@ -284,170 +262,201 @@ public void testConnectivity() {
       }
-      throw new RuntimeException(ex);
-    }
+      if (properties == null) {
+        // If blob properties is null, return null
+        logger.debug("Blob {} not found.", blobId);
+        return null;
+      }
+      return CloudBlobMetadata.fromMap(properties.getMetadata());
+    });
   }
 
   /**
-   * Update the metadata for the specified blob.
+   * Update the metadata for the specified blob asynchronously.
    * @param blobId The {@link BlobId} to update.
    * @param updateFields Map of field names and new values to modify.
    * @param cloudUpdateValidator {@link CloudUpdateValidator} validator for the update passed by the caller.
-   * @return a {@link AzureCloudDestination.UpdateResponse} with the updated metadata.
-   * @throws BlobStorageException if the blob does not exist or an error occurred.
-   * @throws IllegalStateException on request timeout.
+   * @return a CompletableFuture that will eventually contain the {@link UpdateResponse} with the updated metadata.
+   *         If there was an error on ABS side, it will complete exceptionally containing the BlobStorageException.
    */
-  public AzureCloudDestination.UpdateResponse updateBlobMetadata(BlobId blobId, Map<String, Object> updateFields,
-      CloudUpdateValidator cloudUpdateValidator) throws BlobStorageException, StoreException {
+  public CompletableFuture<UpdateResponse> updateBlobMetadataAsync(BlobId blobId, Map<String, Object> updateFields,
+      CloudUpdateValidator cloudUpdateValidator) {
+
+    CompletableFuture<UpdateResponse> resultFuture = new CompletableFuture<>();
     Objects.requireNonNull(blobId, "BlobId cannot be null");
     updateFields.keySet()
         .forEach(field -> Objects.requireNonNull(updateFields.get(field), String.format("%s cannot be null", field)));
 
-    try {
-      Timer.Context storageTimer = azureMetrics.blobUpdateTime.time();
-      try {
-        BlobProperties blobProperties = storageClient.getPropertiesWithResponse(blobId, defaultRequestConditions)
-            .get(requestTimeout.toMillis(), TimeUnit.MILLISECONDS);
-        // Note: above throws 404 exception if blob does not exist.
-        String etag = blobProperties.getETag();
-        Map<String, String> metadata = blobProperties.getMetadata();
+    Timer.Context storageTimer = azureMetrics.blobUpdateTime.time();
 
-        // Validate the sanity of update operation by comparing the fields we want to update against existing metadata
-        // fields in Azure. This can throw StoreExceptions which are propagated to caller.
-        if (!cloudUpdateValidator.validateUpdate(CloudBlobMetadata.fromMap(metadata), blobId, updateFields)) {
-          return new AzureCloudDestination.UpdateResponse(false, metadata);
-        }
+    // TODO: Pass executor instead of default async method.
+    FutureUtils.orTimeout(storageClient.getPropertiesWithResponse(blobId, defaultRequestConditions), requestTimeout)
+        .whenComplete((blobProperties, throwableOnGet) -> {
```

Comment:
I have doubts about this implementation, it seems like we can chain up these logics but we decide to squash them into one `whenComplete` method.  Is there any better way to construct this CompletionStage?

Reply:
Sure. Separated get and update handling into different completion stages and chained them.